core/drivers/versal_trng.c

```diff
@@ -0,0 +1,1142 @@
+}
+
+static void trng_clrset32(vaddr_t addr, uint32_t off, uint32_t mask,
+			  uint32_t val)
+{
+	io_clrsetbits32(addr + off, mask, mask & val);
+}
+
+static void trng_write32_range(const struct versal_trng *trng, uint32_t start,
+			       uint32_t num_regs, const uint8_t *buf)
+{
+	uint32_t off = 0;
+	uint32_t val = 0;
+	size_t cnt = 0;
+	size_t i = 0;
+
+	for (i = 0; i < num_regs; ++i) {
+		if (!buf) {
+			off = start + i * TRNG_BYTES_PER_REG;
+			trng_write32(trng->cfg.addr, off, 0);
+			continue;
+		}
+
+		for (val = 0, cnt = 0; cnt < TRNG_BYTES_PER_REG; ++cnt)
+			val = (val << 8) | buf[i * TRNG_BYTES_PER_REG + cnt];
+
+		off = start + (TRNG_NUM_INIT_REGS - 1 - i) * TRNG_BYTES_PER_REG;
+		trng_write32(trng->cfg.addr, off, val);
+	}
+}
+
+static TEE_Result trng_wait_for_event(vaddr_t addr, uint32_t off,
+				      uint32_t mask, uint32_t event,
+				      uint32_t time_out)
+{
+	uint64_t tref = timeout_init_us(time_out);
+
+	do {
+		if (timeout_elapsed(tref))
+			panic();
+
+	} while ((io_read32(addr + off) & mask) != event);
+
+	return TEE_SUCCESS;
+}
+
+static void trng_soft_reset(const struct versal_trng *trng)
+{
+	trng_clrset32(trng->cfg.addr, TRNG_CTRL, TRNG_CTRL_PRNGSRST_MASK,
+		      TRNG_CTRL_PRNGSRST_MASK);
+	udelay(RESET_DELAY);
+	trng_clrset32(trng->cfg.addr, TRNG_CTRL, TRNG_CTRL_PRNGSRST_MASK, 0);
+}
+
+static void trng_reset(const struct versal_trng *trng)
+{
+	trng_write32(trng->cfg.addr, TRNG_RESET, TRNG_RESET_VAL_MASK);
+	udelay(RESET_DELAY);
+	trng_write32(trng->cfg.addr, TRNG_RESET, 0);
+	trng_soft_reset(trng);
+}
+
+static void trng_hold_reset(const struct versal_trng *trng)
+{
+	trng_clrset32(trng->cfg.addr, TRNG_CTRL,
+		      TRNG_CTRL_PRNGSRST_MASK, TRNG_CTRL_PRNGSRST_MASK);
+	trng_write32(trng->cfg.addr, TRNG_RESET, TRNG_RESET_VAL_MASK);
+	udelay(RESET_DELAY);
+}
+
+static TEE_Result trng_check_seed(uint8_t *entropy, uint32_t len)
+{
+	uint32_t *p = (void *)entropy;
+	size_t i = 0;
+
+	for (i = 0; i < len / sizeof(*p); i++) {
+		if (p[i] == ALL_A_PATTERN_32)
+			return TEE_ERROR_GENERIC;
+
+		if (p[i] == ALL_5_PATTERN_32)
+			return TEE_ERROR_GENERIC;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result trng_collect_random(struct versal_trng *trng, uint8_t *dst,
+				      size_t len)
+{
+	const size_t bursts = len / TRNG_BURST_SIZE;
+	const size_t words = TRNG_BURST_SIZE_BITS / TRNG_REG_SIZE;
+	uint32_t *p = (void *)dst;
+	size_t bcnt = 0;
+	size_t wcnt = 0;
+	size_t index = 0;
+	uint32_t val = 0;
+	bool match = false;
+
+	trng_clrset32(trng->cfg.addr, TRNG_CTRL,
+		      TRNG_CTRL_PRNGSTART_MASK, TRNG_CTRL_PRNGSTART_MASK);
+
+	/* Loop as many times based on len requested. In each burst 128 bits
+	 * are generated, which is reflected in QCNT value of 4 by hardware.
+	 */
+	for (bcnt = 0; bcnt < bursts; bcnt++) {
+		if (trng_wait_for_event(trng->cfg.addr,
+					TRNG_STATUS, TRNG_STATUS_QCNT_MASK,
+					TRNG_MAX_QCNT << TRNG_STATUS_QCNT_SHIFT,
+					TRNG_GENERATE_TIMEOUT)) {
+			EMSG("Time out");
```

Comment:
The `EMSG()` output doesn't show file/line number by default so a bit of context would be nice. How about "versal_trng: Time out"? and same in other places. You may define a macro to add the prefix.

Reply:
Um is this new? AFAIR it always showed the file and line

Reply:
About five years :wink: https://github.com/OP-TEE/optee_os/blob/3.17.0/mk/config.mk#L99-L104
I introduced this mainly for "INFO" messages, and thought "ERROR" level output should probably not include source-code level information either because the end-user should be able to understand the error without checking the code. That being said, we're loosing the function name which often gives some context. Maybe we need a macro to show the driver name like in the Linux kernel?